README.md
# ng-scoreboard-clock

[![Build Status](https://travis-ci.org/zeMirco/ng-scoreboard-clock.png)](https://travis-ci.org/zeMirco/ng-scoreboard-clock)

An angular.js directive to display the current time as seen on sports scoreboards. Check out the [demo](http://zemirco.github.com/ng-scoreboard-clock/).

## How to use

Download `scoreboard.js` and include the script in your `html` file.

```html
<script src="scoreboard.js"></script>
```

Then activate your angular.js app

```html
<html ng-app="myApp">
```

and inject `scoreboard` in your main angular.js JavaScript file

```js
angular.module('myApp', ['scoreboard']);
```

And last but not least place the following code snippet anywhere inside your `html` file.

```html
<clock format="HH:mm"></clock>
```

Be aware of [Internet Explorer](http://docs.angularjs.org/guide/ie) restrictions.

## Configuration

To format the time (you could even add the date) use the format string provided by the
angular.js [date filter](http://docs.angularjs.org/api/ng.filter:date).

Pass them to the directive through the `format` attribute.

```html
<clock format="medium"></clock>
```

## Test

You need to have [karma](http://karma-runner.github.com/0.8/index.html) and [grunt](http://gruntjs.com/)
installed to run the tests.

Run

```bash
$ grunt test
```

## License

Copyright (C) 2013 [Mirco Zeiss](mailto: mirco.zeiss@gmail.com)

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

History.md
0.0.2 / 2013-04-05
==================

 * fix opacity value

0.0.1 / 2013-04-03
==================

 * initial release

test/test.js
describe('scoreboard', function() {

  var elm, scope;

  // load the scoreboard code
  beforeEach(module('scoreboard'));

  beforeEach(inject(function($rootScope, $compile) {
    elm = angular.element(
      '<clock format="HH:mm"></clock>'
    );
    scope = $rootScope;
    $compile(elm)(scope);
    scope.$digest();
  }));

  it('clock should show the current time', function() {
    var date = new Date();
    var hours = ('0' + date.getHours()).slice(-2);
    var minutes = ('0' + date.getMinutes()).slice(-2);
    expect(elm.text()).toBe(hours + ':' + minutes);
  });

  it('clock should allow using custom formats', inject(function($rootScope, $compile) {
    elm = angular.element(
      '<clock format="HH:mm dd-MM-yyyy"></clock>'
    );
    scope = $rootScope;
    $compile(elm)(scope);
    scope.$digest();
    var date = new Date();
    var hours = ('0' + date.getHours()).slice(-2);
    var minutes = ('0' + date.getMinutes()).slice(-2);
    var days = ('0' + date.getDate()).slice(-2);
    var months = ('0' + (date.getMonth()+1)).slice(-2);
    var fullYear = date.getFullYear();

    var res = hours + ':' + minutes + ' ' + days + '-' + months + '-' + fullYear;
    expect(elm.text()).toBe(res);
  }));

});

Gruntfile.js

module.exports = function(grunt) {

  // Project configuration.
  grunt.initConfig({
    pkg: grunt.file.readJSON('package.json'),
    jshint: {
      options: {
        es5: true,
        expr: true
      },
      files: ['Gruntfile.js', 'index.js', 'karma.conf.js', 'test/test.js']
    },
    karma: {
      options: {
        configFile: 'karma.conf.js'
      },
      dev: {
        browsers: ['Chrome']
      },
      travis: {
        browsers: ['PhantomJS'],
        singleRun: true
      }
    }
  });

  // load tasks
  grunt.loadNpmTasks('grunt-contrib-jshint');
  grunt.loadNpmTasks('grunt-karma');

  // register tasks
  grunt.registerTask('hint', ['jshint']);
  grunt.registerTask('ng', ['karma:dev']);
  grunt.registerTask('travis', ['karma:travis']);

};

karma.conf.js
// Karma configuration

// base path, that will be used to resolve files and exclude
basePath = '';

// list of files / patterns to load in the browser
files = [
  JASMINE,
  JASMINE_ADAPTER,

  // libraries
  'example/vendor/jquery-1.9.1.min.js',
  'example/vendor/angular.min.js',
  'example/vendor/angular-mocks.js',

  // our app
  'scoreboard.js',

  // tests
  'test/test.js'
];

// list of files to exclude
exclude = [];

// test results reporter to use
// possible values: 'dots', 'progress', 'junit'
reporters = ['progress'];

// web server port
port = 9876;

// cli runner port
runnerPort = 9100;

// enable / disable colors in the output (reporters and logs)
colors = true;

// level of logging
// possible values: LOG_DISABLE || LOG_ERROR || LOG_WARN || LOG_INFO || LOG_DEBUG
logLevel = LOG_INFO;

// enable / disable watching file and executing tests whenever any file changes
autoWatch = true;

// Start these browsers, currently available:
// - Chrome
// - ChromeCanary
// - Firefox
// - Opera
// - Safari (only Mac)
// - PhantomJS
// - IE (only Windows)
browsers = ['Chrome'];

// If browser does not capture in given timeout [ms], kill it
captureTimeout = 60000;

// Continuous Integration mode
// if true, it capture browsers, run tests and exit
singleRun = false;


scoreboard.js
angular.module('scoreboard', [])
  // inject $timeout and dateFilter service
  .directive('clock', function($timeout, dateFilter) {
    return {
      restrict: 'E',
      template: '<span></span>',
      replace: true,
      link: function(scope, iElem, iAttrs) {
        // save timeoutIds, so that we can cancel the time updates
        var updateTimeoutId;
        var hideTimeoutId;

        var colon = '<span class="colon" style="opacity: 1;">:</span>';

        // schedule update in one second
        var updateLater = function() {
          updateTimeoutId = $timeout(function() {
            updateUi();
          }, 1000);
        };

        // show and hide colon for 500ms, reappear each second
        var hideColon = function() {
          hideTimeoutId = $timeout(function() {
            angular.element('span.colon').css('opacity', '0');
          }, 500);
        };

        // update UI
        var updateUi = function() {
          iElem.html(dateFilter(Date.now(), iAttrs.format).split(':').join(colon));
          hideColon();
          updateLater();
        };

        // listen on DOM destroy and cancel the next UI update
        iElem.bind('$destroy', function() {
          $timeout.cancel(updateTimeoutId);
          $timeout.cancel(hideTimeoutId);
        });

        // start process
        updateUi();

      }
    };
  });

index.js
angular.module('scoreboard', [])
  // inject $timeout and dateFilter service
  .directive('clock', function($timeout, dateFilter) {
    return {
      restrict: 'E',
      template: '<span></span>',
      replace: true,
      link: function(scope, iElem, iAttrs) {
        // save timeoutIds, so that we can cancel the time updates
        var updateTimeoutId;
        var hideTimeoutId;

        var colon = '<span class="colon" style="opacity: 100;">:</span>';

        // schedule update in one second
        var updateLater = function() {
          updateTimeoutId = $timeout(function() {
            updateUi();
          }, 1000);
        };

        // show and hide colon for 500ms, reappear each second
        var hideColon = function() {
          hideTimeoutId = $timeout(function() {
            angular.element('span.colon').css('opacity', '0');
          }, 500);
        };

        // update UI
        var updateUi = function() {
          iElem.html(dateFilter(Date.now(), iAttrs.format).split(':').join(colon));
          hideColon();
          updateLater();
        };

        // listen on DOM destroy and cancel the next UI update
        iElem.bind('$destroy', function() {
          $timeout.cancel(updateTimeoutId);
          $timeout.cancel(hideTimeoutId);
        });

        // start process
        updateUi();

      }
    }
  });